torch/utils/_sympy/value_ranges.py

```diff
@@ -171,10 +179,19 @@ def __init__(self, lower: AllIn, upper: AllIn) -> None:
         object.__setattr__(
             self,
             "is_int",
-            not self.is_bool
-            and (
-                isinstance(lower, (sympy.Integer, NegativeIntInfinity))
-                or isinstance(upper, (sympy.Integer, IntInfinity))
+            (
+                not self.is_bool
+                and (
+                    isinstance(lower, (sympy.Integer, NegativeIntInfinity))
+                    # -oo is a valid lower bound for integers
+                    or lower in (-int_oo, -sympy.oo)
+                )
+                and (
+                    isinstance(upper, (sympy.Integer, IntInfinity))
+                    # oo is a valid upper bound for integers
+                    or upper in (int_oo, sympy.oo)
+                )
+                and (lower, upper) != (-sympy.oo, sympy.oo)  # (-oo, oo) is float
```

Comment:
Hmm. I feel I had misunderstood the original change you had made here, and now I don't understand what is going on here at all.

Previously, we inferred that a VR was integral by testing if you gave us an integer or int infinity. Now, you are also allowing for integral VR if they have -sympy.oo or sympy.oo lower/upper bound. But then naively, this would mean (-sympy.oo, sympy.oo) is integral. Which you have manually patched so that it is not the case. But I don't understand what was wrong with the original logic, this seems to be the right thing to me! No sympy.oo allowed for integer ranges, if one is showing up, that's a bug.

Reply:
> But I don't understand what was wrong with the original logic, this seems to be the right thing to me! No sympy.oo allowed for integer ranges, if one is showing up, that's a bug.

I reverted this and would fix this in a follow-up.

- #131046